Feature/lrf7

whaeck opened the pull request:
LRF7 capability, it handles most of LRF3 options and Reich-Moore LRF7. It can easily be expanded to full R-matrix with relativistic kinematics.

This includes the necessary fromENDF() function to create a Reconstructor object.

whaeck commented on src/resonanceReconstruction.hpp:
```diff
@@ -51,6 +52,36 @@ namespace ENDF = ENDFtk::resonanceParameters;
+using FaradPerMeter = decltype( farad / meter );
+
+// hbar constant in eV s - taken from 2014 CODATA
+constexpr Quantity< ElectronVoltSecond > hbar = 6.582119514e-16 * electronVolt * second;
```
I will change that later. The unit tests result WILL change because of it and it's better to do that separately.

whaeck commented on src/resonanceReconstruction.hpp:
```diff
@@ -9,6 +9,7 @@
+#pragma GCC diagnostic ignored "-Wextra"
 #include <Eigen/Dense>
 #pragma GCC diagnostic pop
 
```
Current Eigen version is 3.3.90 by the way, so MPL2 applies (since Eigen 3.1.1)

whaeck commented on src/resonanceReconstruction/rmatrix/ChannelQuantumNumbers/src/toString.hpp:
```diff
@@ -0,0 +1,23 @@
+  auto toHalfIntegerString = 
+      [] ( const double a )
+         { double half;
+           return std::modf( a, &half ) == 0. ?
```
It's a ternary statement and perfectly legal. The doxygen comments for the toString function explain clearly what this function does.

whaeck commented on src/resonanceReconstruction/rmatrix/ChannelQuantumNumbers/src/toString.hpp:
```diff
@@ -0,0 +1,23 @@
+                      std::to_string( 2 * static_cast<int>( half ) + 1 ) 
+                        + "/2"; };
+
+  return "{" + std::to_string( this->l_ ) + ","
+             + toHalfIntegerString( this->s_ ) + ","
+             + toHalfIntegerString( this->J_ )
+             + ( this->parity_ > 0 ? "+" : "-" ) + "}";
+}
```
Including a complete dependency just for a single format statement isn't worth it to me. I'll change it when we move to C++20 and it's part of the standard :-)

And yes, I know it is an implicit dependency because of ENDFtk and Log.

whaeck commented on src/resonanceReconstruction/rmatrix/src/coh3-coulomb.hpp:
```diff
@@ -0,0 +1,39 @@
+const double HBAR       =  6.582119514e-22; /* Planck's constant/2pi [MeV sec]*/
+const double VLIGHT     =  2.99792458e+23 ; /* light velocty [fm/sec]         */
```
The hbar here is CODATA2014, as is the one in resonanceReconstruction.hpp so it's consistent.

whaeck replied:
The physical constants given here (hbar, c, elementary charge, amu to energy conversion, permittivity) do not appear to be used in the code since I provide eta in the coulomb wave function call. We can probably remove it.

jlconlin commented on src/resonanceReconstruction/rmatrix/src/coh3-coulomb.hpp:
```diff
@@ -1,9 +1,9 @@
-const double COULOMB    =  1.60217733e-19 ; /* J = 1eV                        */
-const double COULOMBSQ  =  2.56697220e-38 ; /* COULOMB*COULOMB                */
-const double PERMITTIV  =  5.60958617e+37 ; /* permittivity [MeV fm /C^2]     */
+//const double HBAR       =  6.582119514e-22; /* Planck's constant/2pi [MeV sec]*/
```
When we have figured out why our tests are not quite correct, we should delete these lines instead of commenting them out.

jlconlin commented on src/resonanceReconstruction/rmatrix/src/coh3-extwave.hpp:
```diff
@@ -192,17 +192,15 @@ inline int omExternalClosed(int lmax, double rho_match, double coulomb, std::com
-}
-
-
+//inline void omCoulombParameter(int zz, double mu, double e, double *eta, double *sig0)
```
Also delete these lines instead of commenting them out.

jlconlin commented on src/resonanceReconstruction/rmatrix/src/makeParticlePairs.hpp:
```diff
@@ -0,0 +1,124 @@
+                     spin == 0.0 ? ( parity >= 0 ? Parity( +1 ) : Parity( -1 ) )
+                                 : ( spin > 0. ? Parity( +1 ) : Parity( -1 ) ) };
+  };
+  auto makeParticlePair = [&] ( const Particle& particle,
```
Is the reason you create this lambda (instead of using the `ParticlePair` constructor directly) so you can convert `Q` to a `Quantity`?

whaeck replied:
You can't zip_with with a constructor, and the Q value needs to be converted to a Quantity, yes.

jlconlin commented on src/resonanceReconstruction/rmatrix/src/possibleChannelTotalAngularMomentumValues.hpp:
```diff
@@ -0,0 +1,56 @@
+ */
+//template < typename Spin >
+inline auto
+getPossibleChannelTotalAngularMomentumValues( const OrbitalAngularMomentum& l,
```
ugh `get`...

jlconlin commented on src/resonanceReconstruction/rmatrix/ChannelQuantumNumbers/test/ChannelQuantumNumbers.test.cpp:
```diff
@@ -22,15 +22,14 @@ SCENARIO( "ChannelQuantumNumbers" ) {
-      REQUIRE( +1 == numbers.parity() );
-      REQUIRE( "{1,1/2,3/2+}" == numbers.toString() );
-    }
+      CHECK( 1 == numbers.orbitalAngularMomentum() );
```
You know this makes me happy

whaeck replied:
Since I'm doing major refactoring, and since I have to go over all tests anyway, I thought I'd make that change at the same time.

jlconlin commented on src/resonanceReconstruction/rmatrix/Channel/src/sommerfeldParameter.hpp:
```diff
@@ -0,0 +1,28 @@
+/**
```
Since you moved this file, you need to re-run `metaconfigure/generate.py`. I'm getting error that the file can't be found.

This will be a nice thing to go away when we have the newer build system.

whaeck replied:
I'm not done yet. I'm going over each class separately and compiling each test separately. No need to commit the same file over and over simply because I rename files.

jlconlin replied:
Fair enough. I was just trying to compile to see if the tests pass. Let me know when I should try again.

whaeck replied:
It's gonna take a while. I hope to be done today.

jlconlin commented:
I converted this Pull Request to a "draft" so that it doesn't accidentally get merged. Cool "new" feature of GitHub.

whaeck commented:
@jlconlin @nathangibson14 
The following changes were made after the last review:
- better comments in the evaluate function
- removed commented out code
- introduced the use of the LMatrixCalculator and RLMatrixCalculator
- removed metaconfigure and moved to new build system
- introduced the use of elementary

whaeck commented:
@jlconlin @nathangibson14 Anybody home?

jlconlin commented:
Let's make sure that we make an issue stating that we know that the CWF are uncertain for some problems. That way we can come back to it and know what we need to do.
